lib/Fhp/Action/SendSEPATransfer.php

```diff
@@ -35,21 +47,52 @@ class SendSEPATransfer extends BaseAction
      */
     public static function create(SEPAAccount $account, string $painMessage): SendSEPATransfer
     {
-        if (preg_match('/xmlns="(.*?)"/', $painMessage, $match) === false) {
+        if (preg_match('/xmlns="(.*?)"/', $painMessage, $xmlns) === false) {
             throw new \InvalidArgumentException('xmlns not found in the PAIN message');
         }
+
+        // Check whether the PAIN message contains the correct message root
+        if (strstr($painMessage, '<CstmrCdtTrfInitn>') === false) {
+            throw new \InvalidArgumentException('Pain message contains a wrong message root');
+        }
+
+        // Check whether the PAIN message contains multiple or only one transfer, should match <NbOfTxs>xx</NbOfTxs> in the XML
+        $nbOfTxs = substr_count($painMessage, '<CdtTrfTxInf>');
+        if ($nbOfTxs === 0) {
+            throw new \InvalidArgumentException('PAIN message contains no credit transfer transactions');
+        }
+
+        $ctrlSum = null;
+        if (preg_match('@<GrpHdr>.*<CtrlSum>(?<ctrlsum>[.0-9]+)</CtrlSum>.*</GrpHdr>@s', $painMessage, $matches) === 1) {
+            $ctrlSum = $matches['ctrlsum'];
+        }
+
+        // Check whether a <PmtInf> block sets <BtchBookg> to false: This means that Single Booking is to be requested
+        $singleBooking = preg_match('@<PmtInf>.*<BtchBookg>false</BtchBookg>.*</PmtInf>@s', $painMessage) === 1;
```

Comment:
Please call this $requestSingleBooking like the member field. It's confusing that it's not the same as $singleTransfer, so better make the name as different-looking as possible.